### PR TITLE
Assert valid_is_prepared

### PR DESCRIPTION
**Pull request by lindsayad**
Refs #24018 #14329

### CODE REVIEW

File: framework/src/base/MeshGeneratorSystem.C [hunk 1]
```diff
@@ -347,6 +347,16 @@ MeshGeneratorSystem::executeMeshGenerators()
       const auto & name = generator->name();
 
       auto current_mesh = generator->generateInternal();
+#ifdef DEBUG
```
Comment (loganharbour) [inline framework/src/base/MeshGeneratorSystem.C:350]:
```suggestion
#ifndef NDEBUG
```

?

File: framework/src/base/MeshGeneratorSystem.C [hunk 2]
```diff
@@ -347,6 +347,16 @@ MeshGeneratorSystem::executeMeshGenerators()
...
+      // Assert that the mesh is either marked as not prepared or if it is marked as prepared,
+      // that it's *actually* prepared
+      if (!_has_bmbb && !MeshTools::valid_is_prepared(*current_mesh))
+        mooseError("'",
+                   name,
+                   "' is marked as prepared but is not actually prepared. Please edit the '",
+                   generator->type(),
+                   "' class to call 'set_isnt_prepared()'");
```
Comment (loganharbour) [inline framework/src/base/MeshGeneratorSystem.C:358]:
```suggestion
        generator->mooseError("The generated mesh is marked as prepared but is not actually prepared. Please edit the '",
                   generator->type(),
                   "' class to call 'set_isnt_prepared()'");
```

**Review by loganharbour — CHANGES_REQUESTED** [review]
while I'm here

File: framework/src/base/MeshGeneratorSystem.C (hunk 1 above)
Comment (lindsayad) [reply framework/src/base/MeshGeneratorSystem.C:350]: this check adds asymptotic slowness. Conventionally in that case we do not add it to `devel`

Comment (loganharbour) [reply framework/src/base/MeshGeneratorSystem.C:350]: okay

**Review by GiudGiud — APPROVED** [review]

**Review by roystgnr — APPROVED** [review]

**Review by loganharbour — APPROVED** [review]

**Pull request merged**